### PR TITLE
use service decoration to find and display links in messages CORE-10191

### PR DESCRIPTION
**Pull request by mmaxim**

### CODE REVIEW

**Comment by mmaxim** [comment]
(Ignoring CI on this, failure in kbfs-server)

File: go/chat/utils/utils.go [hunk 1]
```diff
@@ -2070,6 +2074,60 @@ func DecorateBody(ctx context.Context, body string, offset, length int, decorati
...
+var mailtoRegexp = regexp.MustCompile(`(?:(?:[\w-_.]*)@(?:[\w-]+(?:\.[\w-]+)+))\b`)
+
+func DecorateWithLinks(ctx context.Context, body string) string {
+	var added int
+	offset := 0
+	origBody := body
+
+	allMatches := mailtoRegexp.FindAllStringIndex(ReplaceQuotedSubstrings(body, true), -1)
+	for _, match := range allMatches {
+		bodyMatch := origBody[match[0]:match[1]]
```
Comment (joshblum) [inline go/chat/utils/utils.go]: sanity check match length >= 2 here

File: go/chat/utils/utils.go [hunk 2]
```diff
@@ -2070,6 +2074,60 @@ func DecorateBody(ctx context.Context, body string, offset, length int, decorati
...
+			return true
+		}
+		return false
+	}
+
+	offset = 0
+	origBody = body
+	allMatches = linkRegexp.FindAllStringIndex(ReplaceQuotedSubstrings(body, true), -1)
+	for _, match := range allMatches {
+		bodyMatch := origBody[match[0]:match[1]]
```
Comment (joshblum) [inline go/chat/utils/utils.go]: match length check

File: go/chat/utils/utils.go [hunk 3]
```diff
@@ -2070,6 +2074,60 @@ func DecorateBody(ctx context.Context, body string, offset, length int, decorati
...
+		url := bodyMatch
+		if shouldSkipPrefix(bodyMatch) {
+			continue
+		}
+		if !(strings.HasPrefix(bodyMatch, "http://") || strings.HasPrefix(bodyMatch, "https://")) {
+			url = "http://" + bodyMatch
```
Comment (joshblum) [inline go/chat/utils/utils.go]: why not `https` here?

Comment (mmaxim) [reply go/chat/utils/utils.go]: Safer, will redirect to https if they have it and work if they don't.

**Pull request merged**